### PR TITLE
escape like parameter in cache operations

### PR DESCRIPTION
**Pull request by icewind1991**
For #16580

cc @LukasReschke @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by PVince81** [comment]
Nice!
Is that the only location in the code where we use "LIKE" ? Or did you want to limit the consequences and have a look at the other ones separately ?

**Comment by icewind1991** [comment]
> Is that the only location in the code where we use "LIKE"

Not even close

> Or did you want to limit the consequences and have a look at the other ones separately ?

Correct, small PR's are good PR's

**Comment by LukasReschke** [comment]
```
09:11:11 There were 6 failures:
09:11:11 
09:11:11 1) Test\Files\Cache\Cache::testEscaping with data set #1 ('o%')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
09:11:11 2) Test\Files\Cache\Cache::testEscaping with data set #2 ('oth_r')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
09:11:11 3) Test\Files\Cache\Wrapper\CacheJail::testEscaping with data set #1 ('o%')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
09:11:11 4) Test\Files\Cache\Wrapper\CacheJail::testEscaping with data set #2 ('oth_r')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
09:11:11 5) Test\Files\Cache\Wrapper\CachePermissionsMask::testEscaping with data set #1 ('o%')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
09:11:11 6) Test\Files\Cache\Wrapper\CachePermissionsMask::testEscaping with data set #2 ('oth_r')
09:11:11 Failed asserting that false is true.
09:11:11 
09:11:11 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/cache/cache.php:658
09:11:11 
```

**Comment by LukasReschke** [comment]
> Correct, small PR's are good PR's

:heart_eyes:

File: lib/private/files/cache/cache.php [hunk 1]
```diff
@@ -543,7 +543,7 @@ public function moveFromCache(Cache $sourceCache, $sourcePath, $targetPath) {
 		if ($sourceData['mimetype'] === 'httpd/unix-directory') {
 			//find all child entries
 			$sql = 'SELECT `path`, `fileid` FROM `*PREFIX*filecache` WHERE `storage` = ? AND `path` LIKE ?';
-			$result = \OC_DB::executeAudited($sql, [$sourceStorageId, $sourcePath . '/%']);
+			$result = \OC_DB::executeAudited($sql, [$sourceStorageId, \OC_DB::escapeLikeParameter($sourcePath) . '/%']);
```
Comment (DeepDiver1975) [inline lib/private/files/cache/cache.php]: given the fact that we want to kill OC_DB one day and only use IDatabaseConnection we better try to fix this there.

Furthermore: could doctrines querybuilder help us here?

Comment (icewind1991) [inline lib/private/files/cache/cache.php]: > given the fact that we want to kill OC_DB one day and only use IDatabaseConnection we better try to fix this there.

Was my initial idea, but then we need to change every place where like is used to use `IDatabaseConnection` instead of `OC_DB`, which is probably not something we want for 8.1

> Furthermore: could doctrines querybuilder help us here?

I think so, but again not something for 8.1

**Comment by icewind1991** [comment]
Test failing only on oracle, will look into it

**Comment by LukasReschke** [comment]
Any chance to reanimate this at some point? :smile:

File: lib/private/files/cache/cache.php (hunk 1 above)
Comment (nickvergessen) [inline lib/private/files/cache/cache.php]: We have the query builder now :wink:

**Comment by MorrisJobke** [comment]
@DeepDiver1975 @cmonteroluque #16580 is not in any milestone - this one only was added to 8.2 some months back. I would say, that we postpone this to 9.0

**Comment by cmonteroluque** [comment]
@MorrisJobke you're right... I'm moving to backlog so we actually review, both this and the issue.

**Comment by LukasReschke** [comment]
This may have potential security impact as all other queries as well. Moved back to 9.0.

**Comment by cmonteroluque** [comment]
@LukasReschke yeah, thanks

**Comment by PVince81** [comment]
Some steps to cause trouble:

1. Create a folder "a%b" with some contents
1. Create another folder "aub" with some contents
1. Rename "a%b" to "axb"

**Comment by icewind1991** [comment]
Revived:

- move the cache from `OC_DB` to `IDBConnection`
- add `escapeLikeParameter`  to `IDBConnection` instead of `OC_DB`

**Comment by icewind1991** [comment]
Using a query builder wont save us since we still need to be able to use special characters in LIKE queries so we can't just auto-escape every like query

**Comment by PVince81** [comment]
Thanks a lot @icewind1991. Let's see if CI is kind this time.

File: lib/private/files/cache/cache.php [hunk 2]
```diff
@@ -101,22 +108,22 @@ public function getNumericStorageId() {
 	 *
 	 * the returned cache entry contains at least the following values:
 	 * [
-	 * 		'fileid' => int, the numeric id of a file (see getId)
```
Comment (nickvergessen) [inline lib/private/files/cache/cache.php]: please stop doing such unrelated stuff which just slows down reviews

Comment (icewind1991) [inline lib/private/files/cache/cache.php]: https://github.com/owncloud/core/pull/16604/files?w=1

Comment (nickvergessen) [inline lib/private/files/cache/cache.php]: yeah I know w=1, however since we are mostlikly going to backport this, I'd prefer to keep unrelated changes out of this...

**Comment by PVince81** [comment]
Code looks good, did some quick manual testing and it worked as well.
Smashbox tests passed too.

:+1:

**Comment by MorrisJobke** [comment]
Same here. Tests worked fine :+1:

**Comment by icewind1991** [comment]
@karlitschek backport?

**Comment by PVince81** [comment]
Due to the fact that this could mess up file cache entries and make bogus folders appear, we should probably backport this down to OC 8.0. But also this will need proper regression testing. @rperezb @karlitschek

**Comment by PVince81** [comment]
(it will be an expensive backport as the past code is slightly different)

**Comment by PVince81** [comment]
stable8.2: https://github.com/owncloud/core/issues/20432

It was a simplified backport.

@karlitschek should this simplified backport be backported to 8.1 and 8.0 ? (it only fixes the "move" operation which is the one that caused the known issues)

**Comment by MorrisJobke** [comment]
> @karlitschek should this simplified backport be backported to 8.1 and 8.0 ? (it only fixes the "move" operation which is the one that caused the known issues)

@karlitschek ;)

**Comment by karlitschek** [comment]
@MorrisJobke Agreed. Please backport

**Comment by PVince81** [comment]
@icewind1991 can you create the stable8.1 backport PR ? Thanks

**Comment by icewind1991** [comment]
8.1: https://github.com/owncloud/core/pull/20637
8.0: #20638
